infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/RestoreRedundancyMergeSegments.java

```diff
@@ -1,38 +1,213 @@
 package org.corfudb.infrastructure.orchestrator.actions;
 
-import com.google.common.collect.Sets;
-
-import java.util.Set;
-
-import javax.annotation.Nonnull;
-
+import com.google.common.collect.ImmutableList;
+import lombok.Builder;
+import lombok.Builder.Default;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.log.StreamLog;
+import org.corfudb.infrastructure.log.statetransfer.StateTransferManager;
+import org.corfudb.infrastructure.log.statetransfer.batchprocessor.protocolbatchprocessor.ProtocolBatchProcessor;
+import org.corfudb.infrastructure.log.statetransfer.exceptions.TransferSegmentException;
 import org.corfudb.infrastructure.orchestrator.Action;
+import org.corfudb.infrastructure.redundancy.RedundancyCalculator;
+import org.corfudb.protocols.wireprotocol.Token;
 import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.exceptions.OutrankedException;
+import org.corfudb.runtime.exceptions.QuorumUnreachableException;
+import org.corfudb.runtime.exceptions.RetryExhaustedException;
+import org.corfudb.runtime.exceptions.WrongEpochException;
 import org.corfudb.runtime.view.Layout;
+import org.corfudb.runtime.view.LayoutBuilder;
+import org.corfudb.runtime.view.LayoutManagementView;
+import org.corfudb.util.retry.ExponentialBackoffRetry;
+import org.corfudb.util.retry.IRetry;
+import org.corfudb.util.retry.RetryNeededException;
+
+import javax.annotation.Nonnull;
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.TransferSegment;
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.TransferSegmentStatus.SegmentState.FAILED;
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.TransferSegmentStatus.SegmentState.TRANSFERRED;
 
 /**
- * This action attempts to restore redundancy for all servers across all segments
- * starting from the oldest segment. It then collapses the segments once the set of
- * servers in the 2 oldest subsequent segments are equal.
- * Created by Zeeshan on 2019-02-06.
+ * This action attempts to restore the redundancy for the segments, for which the current node is
+ * missing. It transfers multiple segments at once and opens the segments that have been transferred.
  */
+@Slf4j
+@Getter
+@Builder
 public class RestoreRedundancyMergeSegments extends Action {
```

Comment:
If the fault detector needs to be able to do any merge at any time (Regardless of restore redundancy), why not just have one code path that does that. I dont see a strong reason to spin on that condition in the orchestrator.

Reply:
The point is to centralize layout reconfiguration as much as possible and restoring redundancy is an ad-hoc process doesn't really need to change the layout, so it doesnt make sense to me.

Reply:
Ideally, the node shouldn't even re-add itself to the segment. A prob should detect that range of the address range has been restored and do the appropriate layout changes (i.e. re-add itself to the segment and possibly merge in one shot)